File: 백준/Silver/1920. 수 찾기/수 찾기.py
import sys
input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))
m = int(input())
target_list = list(map(int, input().split()))

a.sort()

def search(st, en, target):
    if st == en:
        if a[st] == target:
            print(1)
        else:
            print(0)
        return
    
    mid = (st + en) // 2
    if a[mid] < target:
        search(mid + 1, en, target)
    else:
        search(st, mid, target)
    
for target in target_list:
    search(0, n-1, target)

File: 백준/Silver/1920. 수 찾기/수 찾기.java

import java.util.*;
import java.io.*;

public class Main {
    public static int n, m;
    public static int[] a;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        a = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            a[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(a);

        m = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < m; i++) {
            int target = Integer.parseInt(st.nextToken());
            binarySearch(target);
        }
    }

    public static void binarySearch(int target) {
        int left = 0;
        int right = n - 1;
        while (left < right) {
            int mid = (left + right) / 2;

            if (target > a[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        if (target == a[left]) {
            System.out.println(1);
        } else {
            System.out.println(0);
        }
    }
}